pkg/healthz/healthz.go

```diff
@@ -0,0 +1,78 @@
+// Package healthz implements healthz handlers.
+package healthz
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	pb "k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1"
+	"sigs.k8s.io/aws-encryption-provider/pkg/plugin"
+)
+
+// NewHandler returns a new healthz handler.
+func NewHandler(srv *plugin.Plugin,
+	cli pb.KeyManagementServiceClient,
+	checkPeriod time.Duration,
+) http.Handler {
+	return &handler{
+		srv:         srv,
+		cli:         cli,
+		checkPeriod: checkPeriod,
+	}
+}
+
+type handler struct {
+	srv         *plugin.Plugin
+	cli         pb.KeyManagementServiceClient
+	checkPeriod time.Duration
+
+	lastMu   sync.Mutex
+	lastErr  error
+	lastTime time.Time
+}
+
+func (hd *handler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	res, err := plugin.Check(hd.cli)
+	if err != nil {
+		rw.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(rw, http.StatusText(http.StatusInternalServerError))
+		zap.L().Error("failed healthz version", zap.Error(err))
+		return
+	}
+	zap.L().Info("passed healthz version", zap.String("version", res))
+
+	err = checkEncrypt(hd)
+	if err != nil {
+		rw.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(rw, err)
+		return
+	}
+
+	rw.WriteHeader(http.StatusOK)
+	fmt.Fprint(rw, http.StatusText(http.StatusOK))
+}
+
+func checkEncrypt(hd *handler) error {
+	hd.lastMu.Lock()
```

Comment:
Do you want a RW lock? Right now only one caller at a time can call this endpoint even if no `hd.svc.Encrypt()` call is performed